### DISCUSSION TITLE
How to solve a one dimensional diffusion problem in each node?

### DISCUSSION THREAD

**Discussion by pengfejeac** [idaholab/moose#26532] [Q&A Modules: General]
Hi there!

I'm working on a three dimensional problem, where it requires solving a one dimensional diffusion problem in each node. I checked some of the previous discussion about nodal kernel and ODE kernel, but none of those seems to be applicable to my case. On the other hand, I tried to customize a nodal kernel for my case, but nodal kernel doesn't have "_grad_phi" and "_phi", I have no idea on how to construct a diffusion problem without "_phi", so I'm stuck. 

Any help will be much appreciated!

**Comment by GiudGiud** [comment]
Hello

Which finite element family are you planning to use for the main variable? And is there only one variable?

Nodal kernels being at nodes, a lot of the families don't have a continuous gradients there and would not work for that purpose. 

Do you have to solve at nodes? Or could you solve three 1D diffusion problem volumetrically instead?

Guillaume

**Reply by pengfejeac** [reply]
Thanks for the reply.  

Any finite element would work. The three dimensional problem has one variable, let's say u, and nodal problem has another variable, let's say v. You can think of u and v as coupled. 

I believe I have to solve at each node, but I'll re-think about this if it's not possible to implement. For one dimensional problem and one dimensional diffusion, I can draw the mesh as one straight line for one problem and a lot of orthogonal line at each node for diffusion problem. But in three dimensional case, I have no idea.

>  Or could you solve three 1D diffusion problem volumetrically instead?

I never thought about this, are there any examples that I can refer to?

**Reply by GiudGiud** [reply]
Actually I have realized it would be easiest to use a multiapp approach here. 

You can define a 1D solve in a separate input file, then distribute a bunch of sub apps, at nodes/quadrature points/element centroids, as you want, and solve them in a distrivuted manner. 

Then after that you can transfer the results of v from the distrivuted 1D solves to the 3D mesh. 

Before the 1D solves you can also transfer the local values of u at every node to the 1D solves

**Reply by GiudGiud** [reply]
Actually we do not have a `NodalMultiApp` or a `NodalPositions` (we can create multiapps on positions: https://mooseframework.inl.gov/docs/PRs/26528/site/syntax/Positions/index.html).
I would be interested to see the latter contributed to MOOSE if you want to code it. 

can you start with solvingthe 1D problems at element centroids otherwise?

**Reply by pengfejeac** [reply]
Sounds good to me. I will start with creating a small three dimensional mesh with a few node and try this distribution of multiapp, which I have never used before. Thanks again for your help.

**Reply by pengfejeac** [reply]
@GiudGiud After some attempt, it seems like distributive 1D subapps is not very ideal. I need not only interactions between each subapps but also taking care of tons of subapp at the same time, which doesn't seems to be a great idea. So I'm thinking about hand-coded residual and Jacobian matrix, diagonal and off diagonal. If I fix the amount of elements in 1D diffusion, then I can obtain and calculate shape functions as well as the gradient.

**Reply by GiudGiud** [reply]
Sub apps are created and run in a distributed manner so having a large number of them should be manageable. 


However you are right that if you need these 1D solves to be communicating in a certain way, this might not work. 

What we can do in terms of communication is create a field in the main app from the 1D solve then send back that field to the 1D apps. 

We can also have transfers between multiapps. This would however mean you need to create multiple multiapps and not just one multiapp with many child app. 

Maybe you could group the 1D solves in multiple multiapps? For example in a grid as I imagine each 1D app only communicates with its neighbors?

Guillaume

**Reply by pengfejeac** [reply]
Wow, that sounds really advanced. Let me clear things up a little bit. 

> We can also have transfers between multiapps. This would however mean you need to create multiple multiapps and not just one multiapp with many child app.

So what you are suggesting is create a "branch" like structure with one Main app, a couple of intermediate app and a ton of sub app, is that correct? And that could be done by grouping? 

Communicate with neighbors sounds like finite difference, but that should work if I can implement this branch.

**Reply by GiudGiud** [reply]
You're right it might be too complicated. I dont think anyone has done this before. 

When you do these 1D solves, are you able to code up the entire solve in an object? If so you can use NodalKernels to perform the 1D solves on every node. 
NodalKernels can even contribute to the residual and Jacobian of the main solve.

**Reply by pengfejeac** [reply]
Sorry about the late reply. 

I'm capable of coding diffusion with certain boundary conditions and calculate entries in Jacobian matrix explicitly. But I'm having trouble understanding the difference between Nodal Kernel, Kernel and Array Kernels in terms of "JxW" term. Will that value be different if I apply Nodal kernel instead of Array kernel? Is there a way that I can avoid that? 

As far as I understand, JxW means Jacobian times weight of Gauss quadrature rules, but I can't see the meaning of it if we are interpreting a node.

**Reply by GiudGiud** [reply]
A nodal kernel is supposed to be a local equation only active on that node. The variable is often very local, with completely independent solves on each node (but you can still pull in information from other nodes if you want)

An array kernel is for a field variable that extends over the entire mesh (or a few subdomains) but also has multiple components. For exmaple, the neutron flux with several energy groups is often an array variable. Or an array of multiple species being advected in a fluid.

**Reply by pengfejeac** [reply]
This is probably the last update. So I managed to code diffusion on each node without using any sub-app (Basically hand-coded residual and Jacobian entries as you suggested) and it takes a huge amount of time to debug. But it eventually worked. I stick with array variable instead of nodal variable, but I guess nodal variable would also work. 

Thanks for your help! :D
